### PR TITLE
Fix Exporter crashing when foreign data objects are null

### PR DESCRIPTION
**Pull request by symball**
I am targetting this branch, because it is the one where bug fixes go

## Changelog

```markdown

### Changed
- current method from the Data Iterator to check whether the referenced document exists

### Fixed
- `PropertyAccessor` crashed if using referenced documents if they are `null`
```

## Subject

The method begins by checking for the presence of a "." in the export field. If found, it is a foreign object so it then checks whether the foreign object really exists or not by trying to get it.

### CODE REVIEW

File: src/Source/DoctrineODMQuerySourceIterator.php [hunk 1]
```diff
@@ -75,9 +75,24 @@ public function current()
         $current = $this->iterator->current();
 
         $data = array();
-
         foreach ($this->propertyPaths as $name => $propertyPath) {
+
+          // Is the property extended
+          if(strpos($propertyPath, '.'))
+          {
```
Comment (core23) [inline src/Source/DoctrineODMQuerySourceIterator.php]: Please run `php-cs-fixer` the code style looks completly wrong

File: src/Source/DoctrineODMQuerySourceIterator.php [hunk 2]
```diff
@@ -75,9 +75,20 @@ public function current()
...
+            $data[$name] = '';
+            // Construct the get method
+            $getMethod = 'get'.ucfirst(strstr($propertyPath, '.', true));
+            // Check whether there is an object
+            if ($current->$getMethod()) {
+                $data[$name] = $this->getValue($this->propertyAccessor->getValue($current, $propertyPath));
+            }
+          } else {
+              $data[$name] = $this->getValue($this->propertyAccessor->getValue($current, $propertyPath));
+          }
```
Comment (greg0ire) [inline src/Source/DoctrineODMQuerySourceIterator.php]: Shouldn't this be recursive? What if there is more than one dot?

File: src/Source/DoctrineODMQuerySourceIterator.php [hunk 3]
```diff
@@ -145,8 +156,14 @@ public function getDateTimeFormat()
      */
     protected function getValue($value)
     {
-        if (is_array($value) || $value instanceof \Traversable) {
-            $value = null;
+        //if value is array or collection, creates string
+        if (is_array($value) or $value instanceof \Traversable) {
```
Comment (greg0ire) [inline src/Source/DoctrineODMQuerySourceIterator.php]: Why change `||` with `or`?

**Comment by greg0ire** [comment]
No need to run php-cs-fixer at level symfony. There is a `.php_cs` file with the rules we want at the root of the project.

File: src/Source/DoctrineODMQuerySourceIterator.php (hunk 3 above)
Comment (symball) [inline src/Source/DoctrineODMQuerySourceIterator.php]: This, will change back. Don't know how or why, doesn't matter

File: src/Source/DoctrineODMQuerySourceIterator.php (hunk 2 above)
Comment (greg0ire) [inline src/Source/DoctrineODMQuerySourceIterator.php]: Ok, great!

**Comment by symball** [comment]
After investigating the propertyAccessor class a bit further, there's a much simpler way of achieving the desired result using the isReadable method. Nice

**Comment by greg0ire** [comment]
Looks way simpler indeed! Can you add some tests or will it be too hard?

**Comment by symball** [comment]
Writing a test to mock a Doctrine Mongo query with referenced documents is above my ability for the moment. Symfony / PHPUnit test is something I am working towards in my learning time though, I just don't see it happening overnight, sorry.
I will follow this to see the end result test however if beaten to it

File: src/Source/DoctrineODMQuerySourceIterator.php [hunk 4]
```diff
@@ -75,9 +75,12 @@ public function current()
         $current = $this->iterator->current();
 
         $data = array();
-
         foreach ($this->propertyPaths as $name => $propertyPath) {
+
```
Comment (core23) [inline src/Source/DoctrineODMQuerySourceIterator.php]: Please remove this empty line

File: src/Source/DoctrineODMQuerySourceIterator.php [hunk 5]
```diff
@@ -75,9 +75,12 @@ public function current()
         $current = $this->iterator->current();
 
         $data = array();
-
```
Comment (core23) [inline src/Source/DoctrineODMQuerySourceIterator.php]: Why?

**Comment by greg0ire** [comment]
No problem, I understand

**Comment by greg0ire** [comment]
Please fix the intention problem shown by styleci

File: src/Source/DoctrineODMQuerySourceIterator.php [hunk 6]
```diff
@@ -75,9 +75,11 @@ public function current()
...
         foreach ($this->propertyPaths as $name => $propertyPath) {
-            $data[$name] = $this->getValue($this->propertyAccessor->getValue($current, $propertyPath));
+            $data[$name] = '';
+            if ($this->propertyAccessor->isReadable($current, $propertyPath)) {
+              $data[$name] = $this->getValue($this->propertyAccessor->getValue($current, $propertyPath));
```
Comment (greg0ire) [inline src/Source/DoctrineODMQuerySourceIterator.php]: Here is the indention problem.

**Comment by core23** [comment]
Why did you remove the `Needs: Tests` tag @greg0ire ?

**Comment by greg0ire** [comment]
b/c @symball explained it would be too hard for him, which I can understand.

**Comment by symball** [comment]
Code styling fixed, thanks for waiting

File: src/Source/DoctrineODMQuerySourceIterator.php (hunk 5 above)
Comment (core23) [inline src/Source/DoctrineODMQuerySourceIterator.php]: Please undo this line

**Comment by OskarStark** [comment]
ping @symball

File: src/Source/DoctrineODMQuerySourceIterator.php [hunk 7]
```diff
@@ -73,15 +73,15 @@ public function __construct(Query $query, array $fields, $dateTimeFormat = 'r')
     public function current()
     {
         $current = $this->iterator->current();
-
+
```
Comment (OskarStark) [inline src/Source/DoctrineODMQuerySourceIterator.php]: please remove whitespace

File: src/Source/DoctrineODMQuerySourceIterator.php [hunk 8]
```diff
@@ -73,15 +73,15 @@ public function __construct(Query $query, array $fields, $dateTimeFormat = 'r')
...
         foreach ($this->propertyPaths as $name => $propertyPath) {
-            $data[$name] = $this->getValue($this->propertyAccessor->getValue($current, $propertyPath));
+            $data[$name] = '';
+            if ($this->propertyAccessor->isReadable($current, $propertyPath)) {
+                $data[$name] = $this->getValue($this->propertyAccessor->getValue($current, $propertyPath));
+            }
         }
-
         $this->query->getDocumentManager()->getUnitOfWork()->detach($current);
-
```
Comment (OskarStark) [inline src/Source/DoctrineODMQuerySourceIterator.php]: please keep this line

**Comment by symball** [comment]
I build the code and make it presentable running php-cs-fixer, try to submit and make the changes many times never seeming to satisfy the damn thing.
What options do you use for php-cs-fixer, honestly? Your documentation on contributing only says:

> php-cs-fixer fix --verbose

Either my php-cs-fixer is broken or that is a lie. I understanding coding to a spec but this is ridiculous and completely put me off wanting to contribute. Please can you sort out the blank lines for yourself?

**Comment by OskarStark** [comment]
You can do it manually, check StyleCI report here: https://styleci.io/analyses/qypj7N

**Comment by greg0ire** [comment]
> Either my php-cs-fixer is broken or that is a lie

Please keep calm, and state your issues… we didn't know you had issues with that, did we. That being said, maybe your php-cs-fixer is outdated? Alternatively, do what @OskarStark suggests and apply the patch from StyleCI, it's guaranteed to work.

**Comment by greg0ire** [comment]
You only have to get rid of the trailing spaces : https://styleci.io/analyses/zdDOOZ

**Comment by symball** [comment]
Genuinely, I am calm; these screenshots are what I have in my editor, both fail with the exact same message on stylci. I'm using Atom on Arch Linux 

screenshot_20170403_182905

screenshot_20170403_183012

**Comment by greg0ire** [comment]
It's trailing spaces, not sure your editor displays them

**Comment by greg0ire** [comment]
https://github.com/atom/whitespace

File: src/Source/DoctrineODMQuerySourceIterator.php [hunk 9]
```diff
@@ -73,15 +73,14 @@ public function __construct(Query $query, array $fields, $dateTimeFormat = 'r')
...
-            $data[$name] = $this->getValue($this->propertyAccessor->getValue($current, $propertyPath));
+            $data[$name] = '';
+            if ($this->propertyAccessor->isReadable($current, $propertyPath)) {
+                $data[$name] = $this->getValue($this->propertyAccessor->getValue($current, $propertyPath));
+            }
         }
-
-        $this->query->getDocumentManager()->getUnitOfWork()->detach($current);
-
+        $this->query->getDocumentManager()->getUnitOfWork()->detach($current);
```
Comment (greg0ire) [inline src/Source/DoctrineODMQuerySourceIterator.php]: Here is the offending line

File: src/Source/DoctrineODMQuerySourceIterator.php (hunk 8 above)
Comment (greg0ire) [inline src/Source/DoctrineODMQuerySourceIterator.php]: @symball keep this line, there must be an empty line before return statements.

**Comment by greg0ire** [comment]
> root committed 3 minutes ago

Do you seriously work as root?

**Comment by symball** [comment]
no i don't, that would be insane. It's just a default for the GIT GUI I and I haven't bothered to change it.

**Comment by greg0ire** [comment]
You just miss a line without any trailing spaces under the line with `detach`

**Comment by sstok** [comment]
https://styleci.io/analyses/X0Wxlw/diff 😈 
`curl -S https://styleci.io/analyses/X0Wxlw/diff | patch -p1`

**Comment by symball** [comment]
I've been working with Linux and web apps for about a decade and have had my fair share of annoying little things to hunt down but, this took the biscuit. Thanks for the patch sstok; instant fix

**Comment by greg0ire** [comment]
> Thanks for the patch sstok; instant fix

> Alternatively, do what @OskarStark suggests and apply the patch from StyleCI, it's guaranteed to work.

Told you :P

> I've been working with Linux and web apps for about a decade and have had my fair share of annoying little things to hunt down but, this took the biscuit.

Me too, and I've decided to strip whitespaces ~5 years ago, you should really do that too, it will save you a lot of headaches.

File: src/Source/DoctrineODMQuerySourceIterator.php [hunk 10]
```diff
@@ -145,8 +145,14 @@ public function getDateTimeFormat()
...
     {
+        //if value is array or collection, creates string
         if (is_array($value) || $value instanceof \Traversable) {
-            $value = null;
+            $result = array();
+            foreach ($value as $item) {
+                $result[] = $this->getValue($item);
+            }
+            $value = implode(', ', $result);
+            //formated datetime output
```
Comment (greg0ire) [inline src/Source/DoctrineODMQuerySourceIterator.php]: "formated" => "formatted"

Comment (greg0ire) [inline src/Source/DoctrineODMQuerySourceIterator.php]: Aside from that, LGTM

**Comment by OskarStark** [comment]
I would like to have a test for this change and can you please check, if we need to do this change in `DoctrineORMQuerySourceIterator`, not just ODM?

**Comment by OskarStark** [comment]
The repository is deleted, look like this PR is dead @core23

**Pull request closed without merging**